File: javascript/core/src/engine/verbs/difference.ts
/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */
import type { TableStore } from '../../index.js'
import type { SetOperationArgs,Step, TableContainer } from '../../types.js'
import { SetOp } from '../../types.js'
import { SetOperationNode } from '../factories.js'
import { setWithStore } from '../util/index.js'
/**
 * Executes an arquero table difference.
 */
export async function difference(
	step: Step<SetOperationArgs>,
	store: TableStore,
): Promise<TableContainer> {
	return setWithStore(step, store, SetOp.Difference)
}

export function differenceNode(id: string): SetOperationNode {
	return new SetOperationNode(id, SetOp.Difference)
}

File: python/data-wrangling-components/data_wrangling_components/engine/verbs/difference.py
#
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project.
#

import pandas as pd

from data_wrangling_components.engine.verbs.verb_input import VerbInput
from data_wrangling_components.table_store import TableContainer


def difference(input: VerbInput):
    input_table = input.get_input()
    others = input.get_others()
    others = pd.concat(others)

    output = input_table.merge(others, how="left", indicator=True)
    output = output[output["_merge"] == "left_only"]
    output = output.drop("_merge", axis=1)

    return TableContainer(table=output)
